gdal/swig/python/gdal-utils/osgeo_utils/gdal_coord_val.py

```diff
@@ -0,0 +1,268 @@
+    if ds is None:
+        raise Exception(f'Could not open {filename}.')
+    if not isinstance(x_arr, NumpyCompatibleArray.__args__):
+        x_arr = (x_arr,)
+    if not isinstance(y_arr, NumpyCompatibleArray.__args__):
+        y_arr = (y_arr,)
+    if len(x_arr) != len(y_arr):
+        raise Exception(f'len(x)={len(x_arr)} should be the same as len(y)={len(y_arr)}')
+    count = len(x_arr)
+    if not isinstance(x_arr, np.ndarray):
+        x_arr = np.array(x_arr)
+    if not isinstance(y_arr, np.ndarray):
+        y_arr = np.array(y_arr)
+
+    if coord_type == CoordType.SameAsDS_SRS:
+        coord_type = None
+
+    # Build Spatial Reference object based on coordinate system, fetched from the opened dataset
+    if coord_type != CoordType.PixelLine:
+        if coord_type is not None:
+            source_srs = ds.GetSpatialRef()
+            # source_srs = osr.SpatialReference()
+            # source_srs.ImportFromWkt(ds.GetProjection())
+            if isinstance(coord_type, osr.SpatialReference):
+                target_srs = coord_type
+            elif isinstance(coord_type, int):
+                target_srs = osr.SpatialReference()
+                target_srs.ImportFromEPSG(coord_type)
+            elif isinstance(coord_type, str):
+                target_srs = osr.SpatialReference()
+                target_srs.ImportFromProj4(coord_type)
+            elif coord_type == CoordType.SameAsDS_SRS_GeogCS:
+                target_srs = source_srs.CloneGeogCS()
+            else:
+                raise Exception(f'Unknown Coordinate Transformation or Target SRS '
+                                f'{coord_type}')
+            coord_type = None if source_srs.IsSame(target_srs) \
+                else osr.CoordinateTransformation(source_srs, target_srs)
+            if coord_type is not None:
+                for i in range(count):
+                    x_arr[i], y_arr[i], _z = coord_type.TransformPoint(x_arr[i], y_arr[i])
+                # coordinate_transformation_or_xy_srs.TransformPoints(points)
+
+        # Read geotransform matrix and calculate corresponding pixel coordinates
+        geomatrix = ds.GetGeoTransform()
+        inv_geometrix = gdal.InvGeoTransform(geomatrix)
+        if inv_geometrix is None:
+            raise Exception("Failed InvGeoTransform()")
+
+        x_arr, y_arr = \
+            (inv_geometrix[0] + inv_geometrix[1] * x_arr + inv_geometrix[2] * y_arr), \
+            (inv_geometrix[3] + inv_geometrix[4] * x_arr + inv_geometrix[5] * y_arr)
+
+    if print_xy:
+        print(f'x={x_arr}, y={y_arr}')
+
+    bands = get_bands(ds, band_nums)
+    shape = (len(bands), count)
+    dt, np_dt = GDALTypeCodeAndNumericTypeCodeFromDataSet(ds)
+    nodata_value = [(band.GetNoDataValue() or 0) for band in bands]
+    nodata_value = np.asarray(nodata_value, dtype=np_dt)
+    results = np.empty(shape=shape, dtype=np_dt)
+
+    for i, (x, y) in enumerate(zip(x_arr, y_arr)):
+        if coord_is_pixel_center:
```

Comment:
IMHO, this setting and what it does are useless. Just cast to int / floor should do the right thing.

Reply:
It might be redundant if you are dealing with a sampling algorithms, but what about other samplers like average? doesn't RasterIO support these?

Reply:
> It might be redundant if you are dealing with a sampling algorithms, but what about other samplers like average? doesn't RasterIO support these?

yes, but we'd better keep with the area-is-pixel GDAL is using to avoid confusion (or more exactly to let code outside GDAL be confused :-) and not import the confusion in our code base). What could be useful is to have some option where you'd compute bilinear interpolation from the surrounding 4 postings of the specified point. Using RasterIO resampling for that could be a bit cumbersome as it is designed to resample something with an extent not a point

Reply:
Is there a Python API to sample point(s) values from a raster band that accepts int or float pixel/line, except for RasterIO?

Reply:
> Is there a Python API to sample point(s) values from a raster band that accepts int or float pixel/line, except for RasterIO?

no. Just thinking that ReadAsArray(float_x, float_y, 1, 1, resample_alg=gdal.GRIORA_Bilinear) might do the right thing. You should check that with (float_x % 1), (float_y % 1) slightly above (0,0), it gets the same value than sampling at int(float_x), int(float_y), and that at (float_x %1), (float_y % 1) slightly below (1,1), it gets the same value than at int(float_x)+1,int(float_y)+1

Reply:
> ReadAsArray(float_x, float_y, 1, 1, resample_alg=gdal.GRIORA_Bilinear)

That's what I've used, from an error message I've realized it is using RasterIO eventually, right?
I'll run the tests you advised.

Reply:
@rouault I've ported all the functionality from `gdallocationinfo.cpp` into `gdallocationinfo.py` (except -xml, -lifonly).
The Python version now supports also float pixel/line for bilinear interpolation and multiple X Y points. as well as more SRS options.

As I want to support both Nearest Neighbor, Bilinear and both int, float arrays - I still can't figure out how to avoid using the workaround of the pixel_offset + line_offset that I've used (coord_is_pixel_center from before).
It seems that Bilinear interprets pixel=0.5 as "the midway between the first two pixels" and not "the middle of the first pixel". Now when I use inverse geotransform on the coordinate of the middle of the first pixel I get 0.5, which doesn't match with what Bilinear expects.
Do you have a better idea how to settle this issue?

Reply:
`x[idx], y[idx], _z = ct.TransformPoint(x[idx], y[idx])` - is there a better way? `ct.TransformPoints` seem to expect something else...

Reply:
After further thinking about the offet issue I think I understand it, RasterIO expects the corner of the extent of the area, thus for a pixel is the corner of the pixel, thus I needed to deduct 0.5 from the pixel and from the line in any case.
Right? 

I think we need a simple core function `PixelsIO` that accepts pixel/line buffers of some pixel centers and returns/sets the band data for the requested pixels.

There could be a faster/simpler version with int offsets for nearest neighbor.

What do you think?

Reply:
GDAL's convention for georeferenced coordinates is that they correspond to the top-left corner of a pixel, not its center. I don't think we should modify this well established convention.
This PR is a bit too big to be reviewed. It should be split in one per utility.

Reply:
OK :)
https://github.com/OSGeo/gdal/pull/3678
https://github.com/OSGeo/gdal/pull/3684
https://github.com/OSGeo/gdal/pull/3685